### PR TITLE
[Mobile Payments] Show Onboarding if not complete when tapping `Collect Payment` on Order Details

### PR DESCRIPTION
**Pull request by joshheald**
Part of: 
- #6607 

### Description

On opening Order Details, this checks the onboarding state for In-Person-Payments.

When the merchant taps `Collect Payment`, it shows onboarding if required.

If onboarding is displayed, when the state becomes `.complete`, the onboarding flow is dismissed and the card payment process started.

There is code added which lays the groundwork for checking the reader connection status first, but at present, only the onboarding state is used.

#### Note on testing

This commit removes the check for a card-present eligible payment gateway account which used to be part of the decision whether to show the `Collect Payment` button on Order Details or not. I did this to make the changes easier to test, and to be closer in line with Android and expected direction of the changes to the `Collect Payment` button.

To restore the previous behaviour, we can revert 3f21f9bb if desired, and we can decide on that at the end of the work towards #6607.

### Testing instructions

Using a store with both Stripe and WCPay plugins installed and activated

1. On the orders tab, tap + and use Create Order
2. Add a product, and tap next to go through to Order Details
3. Tap `Collect Payment`
4. Observe that you see the "Multiple Conflicting Plugins" onboarding screen
5. Use the `Manage Plugins` CTA and disable one of the payment plugins
6. Tap `Done`
7. Observe that you see an onboarding loading screen, which then dismisses and returns to the Order Details
8. Observe that the Collect Payment flow starts.
9. Connect a reader and collect payment.

### Screenshots

https://user-images.githubusercontent.com/2472348/166256431-386246be-d092-40b3-8c58-36d3ad7d2b47.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /// Checks whether there is a reader connected (implying that we're ready to accept payments)
+    /// If there's not, checks whether `CardPresentOnboardingState` is `.completed`
+    ///
+    func checkCardPaymentReadiness() {
+        onboardingUseCase.refresh()
+        let readerConnected = CardPresentPaymentAction.checkCardReaderConnected { connectPublisher in
+            // TODO: Use readerConnectedReadiness to preempt the onboarding readiness check.
+            // This requires a refactor of `checkCardReaderConnected` to emit an event for a connected reader.
```
Comment (ctarda) [inline WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift:41]: Would it make sense to log a ticket for the refactor suggested and link to it directly?

File: WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import Foundation
+import Combine
+import Yosemite
+
+final class CardPaymentReadinessUseCase {
```
Comment (ctarda) [inline WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift:5]: I was going to suggest naming this CardPresentPaymentsReadinessUseCase, but it's too long. 

On the one hand, we have moved to word the feature In-Person Payments publicly, on the other, all our code still calls it CardPresentPayments.

So, maybe IPPReadinessUseCase? I don't know. Definitely not a blocker.

File: WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            // This requires a refactor of `checkCardReaderConnected` to emit an event for a connected reader.
+            let readerConnectedReadiness = connectPublisher.map { _ -> CardPaymentReadiness in
+                return CardPaymentReadiness.loading
+            }
+
+            let onboardingReadiness = self.onboardingUseCase.statePublisher
+                .compactMap({ state -> CardPaymentReadiness? in
+                    switch state {
+                    case .loading:
+                        // Ignoring intermediate loading steps simplifies the logic
```
Comment (ctarda) [inline WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift:51]: Just out of curiosity, what intermediate loading steps would we be ignoring?

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift [hunk 4]
```diff
@@ -49,7 +49,6 @@ final class OrderDetailsDataSource: NSObject {
             isOrderCurrencyEligibleForCardPayment() &&
             isOrderStatusEligibleForCardPayment() &&
             isOrderPaymentMethodEligibleForCardPayment() &&
-            hasCardPresentEligiblePaymentGatewayAccount() &&
```
Comment (ctarda) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: As you mentioned in the PR description, it might be a good idea to revert this change before merging

**Review by ctarda — APPROVED** [review]
I tested on an eligible order created days ago, just to cover that case as well, and it works as advertised.

https://user-images.githubusercontent.com/2722505/166302701-e7589271-0154-4f2b-8711-f2d5d2ce3a5a.mov

One comment about naming, and one request to revert any changes in the logic presenting the "Collect Payment" button, but other than that, :shipit:

File: WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift:51]: The `loading` here matches the times when the onboarding flow shows "Connecting to your account" screens, e.g. at 0:55 in my video, after dismissing the wp-admin modal.

Since we're doing this readiness check in the background initially, we don't need to display the loading state. We also respond to the state when the onboarding flow is on screen, to dismiss it when the state moves to `.complete`.

When the onboarding state is loading, the readiness has not changed. It _might_ change as a result of the loading, but retaining the existing state until it _does_ change makes the use case's output clearer and easier to respond to.

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift (hunk 4 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: Sure. I'll just keep adding it back with each PR

File: WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift:5]: Yeah can do. I was trying to avoid using CPP and IPP abbreviations, but your right to call this out as "Card Payment" isn't specific enough for what this checks. 

On balance, even though it's long, I prefer `CardPresentPaymentsReadinessUseCase` for the specificity and consistency with the rest of our code. I'll change it in the following PR.

File: WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewModels/Order Details/CardPaymentReadinessUseCase.swift:41]: Added:
- #6766
